File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/factory_method/lab/garden.cpp
// garden.cpp

#include "conventional_tomato.h"
#include "dp_exception.h"
#include "organic_tomato.h"
#include "garden.h"
#include "tomato.h"

garden::garden(tomato* a_tomato_type)
{
	set_tomato_type(a_tomato_type);
}

void garden::set_tomato_type(tomato* a_tomato_type)
{
	if (a_tomato_type)
	{
		my_tomato_type = a_tomato_type;
	}
	else 
	{
		my_tomato_type = NULL;
	}
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/factory_method/lab/Garden.java
// Garden.java

public class Garden
{
	public Garden(int aTomatoType)
	{
		setTomatoType(aTomatoType);
	}

	public void setTomatoType(int aTomatoType)
	{
		if (aTomatoType != 1 && aTomatoType != 2)
		{
			throw new DpException("void setTomatoType(int aTomatoType) -- A Garden must have a valid Tomato type.");
		}

		myTomatoType = aTomatoType;
	}

	public Tomato makeTomato()
	{
		if (myTomatoType == 1)
		{
			return new ConventionalTomato(1, 2);
		}
		else if (myTomatoType == 2)
		{
			return new OrganicTomato(1, 2);
		}
		else
		{
			throw new DpException("Tomato makeTomato() -- Oops, should never reach this point.");
		}
	}

	private int myTomatoType;
}
